lib/ipv4/ipv4.ml

```diff
@@ -97,61 +97,66 @@ module Make(Ethif: V1_LWT.ETHIF) (Arpv4 : V1_LWT.ARP) = struct
   end
 
   let adjust_output_header ~dmac ~tlen frame =
-    Wire_structs.set_ethernet_dst dmac 0 frame;
-    let buf = Cstruct.sub frame Wire_structs.sizeof_ethernet Wire_structs.Ipv4_wire.sizeof_ipv4 in
+    let open Ipv4_wire in
+    Ethif_wire.set_ethernet_dst dmac 0 frame;
+    let buf = Cstruct.sub frame Ethif_wire.sizeof_ethernet sizeof_ipv4 in
     (* Set the mutable values in the ipv4 header *)
-    Wire_structs.Ipv4_wire.set_ipv4_len buf tlen;
-    Wire_structs.Ipv4_wire.set_ipv4_id buf (Random.int 65535); (* TODO *)
-    Wire_structs.Ipv4_wire.set_ipv4_csum buf 0;
+    set_ipv4_len buf tlen;
+    set_ipv4_id buf (Random.int 65535); (* TODO *)
+    set_ipv4_csum buf 0;
     let checksum = Tcpip_checksum.ones_complement buf in
-    Wire_structs.Ipv4_wire.set_ipv4_csum buf checksum
+    set_ipv4_csum buf checksum
 
   let allocate_frame t ~dst ~proto =
+    let open Ipv4_wire in
     let ethernet_frame = Io_page.to_cstruct (Io_page.get 1) in
-    let smac = Macaddr.to_bytes (Ethif.mac t.ethif) in
-    Wire_structs.set_ethernet_src smac 0 ethernet_frame;
-    Wire_structs.set_ethernet_ethertype ethernet_frame 0x0800;
-    let buf = Cstruct.shift ethernet_frame Wire_structs.sizeof_ethernet in
-    (* Write the constant IPv4 header fields *)
-    Wire_structs.Ipv4_wire.set_ipv4_hlen_version buf ((4 lsl 4) + (5)); (* TODO options *)
-    Wire_structs.Ipv4_wire.set_ipv4_tos buf 0;
-    Wire_structs.Ipv4_wire.set_ipv4_off buf 0; (* TODO fragmentation *)
-    Wire_structs.Ipv4_wire.set_ipv4_ttl buf 38; (* TODO *)
-    let proto = Wire_structs.Ipv4_wire.protocol_to_int proto in
-    Wire_structs.Ipv4_wire.set_ipv4_proto buf proto;
-    Wire_structs.Ipv4_wire.set_ipv4_src buf (Ipaddr.V4.to_int32 t.ip);
-    Wire_structs.Ipv4_wire.set_ipv4_dst buf (Ipaddr.V4.to_int32 dst);
-    let len = Wire_structs.sizeof_ethernet + Wire_structs.Ipv4_wire.sizeof_ipv4 in
-    (ethernet_frame, len)
+    let len = Ethif_wire.sizeof_ethernet + sizeof_ipv4 in
+    match Ethif_print.print_ethif_header ~buf:ethernet_frame
+      ~ethertype:Ethif_wire.IPv4 ~src_mac:(Ethif.mac t.ethif)
+      ~dst_mac:(Macaddr.broadcast) with
+    | Error s -> 
+      Log.info (fun f -> f "IP.allocate_frame: could not print ethernet header: %s" s);
+      (ethernet_frame, len)
+    | Ok () ->
+      let buf = Cstruct.shift ethernet_frame Ethif_wire.sizeof_ethernet in
+      (* TODO: why 38 for TTL? *)
+      match Ipv4_print.print_ipv4_header ~buf ~src:t.ip ~dst ~proto ~ttl:38 with
+      | Error s ->
+        Log.info (fun f -> f "IP.allocate_frame: could not print IPv4 header: %s" s);
+        (ethernet_frame, len)
```

Comment:
That's probably better than returning an invalid buffer (and I guess there's a bug if we get here anyway).